server/routerlicious/packages/routerlicious-base/src/riddler/tenantManager.ts

```diff
@@ -202,36 +229,121 @@ export class TenantManager {
      */
-    public async refreshTenantKey(tenantId: string): Promise<string> {
-        const db = await this.mongoManager.getDatabase();
-        const collection = db.collection<ITenantDocument>(this.collectionName);
+    public async refreshTenantKey(tenantId: string, keyName: string): Promise<ITenantKeys> {
+        if (keyName !== KeyName.key1 && keyName !== KeyName.key2) {
+            return Promise.reject(new Error("Key name must be either key1 or key2."));
+        }
+
+        const tenantDocument = await this.getTenantDocument(tenantId, false);
 
-        const tenantKey = this.generateTenantKey();
-        const encryptedTenantKey = this.secretManager.encryptSecret(tenantKey);
-        if (encryptedTenantKey == null) {
+        const newTenantKey = this.generateTenantKey();
+        const encryptedNewTenantKey = this.secretManager.encryptSecret(newTenantKey);
+        if (encryptedNewTenantKey == null) {
             winston.error("Tenant key encryption failed.");
             Lumberjack.error("Tenant key encryption failed.", { [BaseTelemetryProperties.tenantId]: tenantId });
             return Promise.reject(new Error("Tenant key encryption failed."));
         }
 
-        await collection.update({ _id: tenantId }, { key: encryptedTenantKey }, null);
+        const tenantKeys = await this.getUpdatedTenantKeys(
+            tenantDocument.key,
+            tenantDocument.secondaryKey,
+            keyName,
+            newTenantKey,
+            tenantId);
 
-        return tenantKey;
+        const updateKey = keyName === KeyName.key2
+                            ? { secondaryKey: encryptedNewTenantKey }
+                            : { key: encryptedNewTenantKey };
+        const db = await this.mongoManager.getDatabase();
+        const collection = db.collection<ITenantDocument>(this.collectionName);
+        await collection.update({ _id: tenantId }, updateKey, null);
+
+        return tenantKeys;
+    }
+
+    /**
+     * Gets updated 2 tenant keys after refresh.
+     */
+    private async getUpdatedTenantKeys(
+        key1: string,
+        key2: string,
+        keyName: string,
+        newTenantKey: string,
+        tenantId: string,
+    ):Promise<ITenantKeys> {
+        // if key2 is to be refreshed
+        if (keyName === KeyName.key2) {
+            const decryptedTenantKey1 = this.secretManager.decryptSecret(key1);
+            if (decryptedTenantKey1 == null) {
+                winston.error("Tenant key1 decryption failed.");
+                Lumberjack.error("Tenant key1 decryption failed.", { [BaseTelemetryProperties.tenantId]: tenantId });
+                return Promise.reject(new Error("Tenant key1 decryption failed."));
+            }
+            return {
+                key1: decryptedTenantKey1,
+                key2: newTenantKey,
+            };
+        }
+
+        // below is if key1 is to be refreshed
+        // if key2 doesn't exist, no need to decrypt
+        if (!key2) {
```

Comment:
Yes it's possible keyname is key2, but key2 is missing. Currently all the existing resources have empty key2. After the changes are deployed, customers can refresh the empty key2 to get new key2. The purpose to support 2 keys is to support key rotation that customers can have 2 keys and they can switch to the other key before refreshing the current key.

Reply:
Got it. But in this case where `key2` is missing, the response below is 
```ts
return {
    key1: newTenantKey,
    key2: "",
};
```

So even though `keyName` was `key2`, we are effectively refreshing `key1`. Is that desired? Are we doing that because this is a backwards compatibility scenario? For instance, to not cause errors when a `key2` refresh is requested, but the customer still only has `key1`.

Reply:
Oh.. Here the keyname is key1. If the keyname is key2, we already return the response on line 312. From line 327 to the end of the function we are talking about keyname === key1.

Reply:
You are right! I got confused and now I realize my comment does not make sense. Sorry about that and thanks for clarifying!